File: src/main/java/com/hazza/LeetCodeByJava/SegmentOne/No64MinimumPathSum.java
package com.hazza.LeetCodeByJava.SegmentOne;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/minimum-path-sum/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-19
 * Time: 9:17 AM
 */
public class No64MinimumPathSum {
    public int minPathSum(int[][] grid) {
        int row = grid.length, col = grid[0].length;
        int[][] dp = new int[row][col];
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (i != 0 && j != 0)
                    dp[i][j] = Math.min(grid[i][j] + dp[i - 1][j], grid[i][j] + dp[i][j - 1]);
                else if (i != 0)
                    dp[i][j] = grid[i][j] + dp[i - 1][j];
                else if (j != 0)
                    dp[i][j] = grid[i][j] + dp[i][j - 1];
                else
                    dp[i][j] = grid[i][j];
            }
        }

        return dp[row - 1][col - 1];
    }
}

File: src/main/python/com/hazza/leetcode_by_python/SegmentTwo/No64_Minimum_Path_Sum.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/6/14 19:58
# https://leetcode.com/problems/minimum-path-sum/
import sys
from typing import List


class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dp = [sys.maxsize] * (n + 1)

        dp[1] = 0
        for i in range(m):
            for j in range(n):
               dp[j + 1] = min(dp[j + 1], dp[j]) + grid[i][j]

        return dp[n]


if __name__ == '__main__':
    solution = Solution()

    input = [
        [1, 3, 1],
        [1, 5, 1],
        [4, 2, 1]
    ]
    print(solution.minPathSum(input))
    assert solution.minPathSum(input) == 7
